### PR TITLE
Feat/partial rollouts

### PR DESCRIPTION
**Pull request by SergiX44**
- Added new partial rollout feature

### CODE REVIEW

**Review by gaedelprete — APPROVED** [review]
👏

File: app/Repositories/BuildRepository.php [hunk 1]
```diff
@@ -242,4 +249,53 @@ public function storeArtifact(
...
+     * @param  string  $deviceId
+     * @return bool
+     */
+    public function isDeviceInRolloutRange(Build $build, string $deviceId): bool
+    {
+        // only devices that contacted the backend
+        // after this date are considered active
+        $threshold = now()
+            ->subDays(config('uppy.active_device_threshold'))
+            ->toDateTimeString();
```
Comment (ilpes) [inline app/Repositories/BuildRepository.php:264]: It's not clear to me the reason for this. What do these 30 days represent?

File: tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php [hunk 2]
```diff
@@ -154,4 +155,90 @@ public function it_allows_null_device_id()
         $response->assertNotFound();
     }
 
+    /**
+     * @test
+     */
+    public function it_return_the_update_when_device_id_is_in_partial_rollout_range(): void
```
Comment (ilpes) [inline tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php:161]:
```suggestion
    public function it_returns_the_update_when_device_id_is_in_partial_rollout_range(): void
```

File: tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php [hunk 3]
```diff
@@ -154,4 +155,90 @@ public function it_allows_null_device_id()
...
+            }
+        }
+
+        $this->assertSame(10, $updateAvailable);
+    }
+
+    /**
+     * @test
+     */
+    public function it_return_the_last_build_if_device_is_not_in_the_partial_rollout_range(): void
```
Comment (ilpes) [inline tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php:202]: Can you please add a test without `$build1`. I'm worried that that `while...do` may potentially cause an infinite loop

**Review by ilpes — CHANGES_REQUESTED** [review]

File: tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php (hunk 3 above)
Comment (SergiX44) [reply tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php:202]: Added another test, the issue was not the infinite loop but a corner case when the rollout percentage is zero (which doesn't make much sense but still a broken behaviour) which will return that the device is always in the range.  Fixed in https://github.com/everli/uppy/pull/8/commits/853108fb3be394d4edaf40d79277865e1bce67f9

File: app/Repositories/BuildRepository.php (hunk 1 above)
Comment (SergiX44) [reply app/Repositories/BuildRepository.php:264]: It's to try to identify in some way the active users of an application: with this pr, we update every time the timestamp of the last time a device has done the check for updates. in this way, we can somehow identify the active devices of each application, in order to make the partial rollout a little more effective. If for example a device hasn't done the check for updates since 3 months, it means that the application hasn't been opened since 3 months, therefore uninstalled or unused, so I think it makes sense to exclude that device from the partial rollout range.

**Review by ilpes — APPROVED** [review]
